### DISCUSSION TITLE
Repeatable fields in eForms Notice Editor (Physical model)

### DISCUSSION THREAD

**Discussion by AlexAS81** [OP-TED/eForms-SDK#328] [Q&A] [answered]
We are testing the "Generate XML" functionality of the eforms notice editor and we see that the repeatable elements do appear in the visual model, but the same does not happen in the physical model.

For example for the field "Buyer":

![image](https://user-images.githubusercontent.com/115476802/217844676-8a9014e7-78b4-41b4-865d-6cf01a6cae6a.png)

Visual model:

`        "contentId": "GR-ContractingAuthority",
          "contentType": "group",
          "contentCount": "1",
          "children": [
            {
              "contentId": "OPT-300-Procedure-Buyer",
              "contentType": "field",
              "contentCount": "1",
              "visType": "field",
              "value": "ORG-0001"
            },
            {
              "contentId": "BT-11-Procedure-Buyer",
              "contentType": "field",
              "contentCount": "1",
              "visType": "field",
              "value": "body-pl"
            },
            {
              "contentId": "BT-10-Procedure-Buyer",
              "contentType": "field",
              "contentCount": "1",
              "visType": "field",
              "value": "defence"
            },
            {
              "contentId": "BT-508-Procedure-Buyer",
              "contentType": "field",
              "contentCount": "1",
              "visType": "field",
              "value": "1111"
            },
            {
              "contentId": "GR-Procedure-SProvider",
              "contentType": "group",
              "contentCount": "1",
              "children": [
                {
                  "contentId": "OPT-300-Procedure-SProvider",
                  "contentType": "field",
                  "contentCount": "1",
                  "visType": "field",
                  "value": "ORG-0001"
                },
                {
                  "contentId": "OPT-030-Procedure-SProvider",
                  "contentType": "field",
                  "contentCount": "1",
                  "visType": "field",
                  "value": "ted-esen"
                }
              ],
              "visType": "non-field",
              "visNodeId": "ND-ServiceProviderParty"
            },
            {
              "contentId": "GR-Procedure-SProvider",
              "contentType": "group",
              "contentCount": "2",
              "children": [
                {
                  "contentId": "OPT-300-Procedure-SProvider",
                  "contentType": "field",
                  "contentCount": "2",
                  "visType": "field",
                  "value": "ORG-0001"
                },
                {
                  "contentId": "OPT-030-Procedure-SProvider",
                  "contentType": "field",
                  "contentCount": "2",
                  "visType": "field",
                  "value": "serv-prov"
                }
              ],
              "visType": "non-field",
              "visNodeId": "ND-ServiceProviderParty"
            }
          ],
          "visType": "non-field",
          "visNodeId": "ND-ContractingParty"
        },
        {
          "contentId": "GR-ContractingAuthority",
          "contentType": "group",
          "contentCount": "2",
          "children": [
            {
              "contentId": "OPT-300-Procedure-Buyer",
              "contentType": "field",
              "contentCount": "2",
              "visType": "field",
              "value": "ORG-0001"
            },
            {
              "contentId": "BT-11-Procedure-Buyer",
              "contentType": "field",
              "contentCount": "2",
              "visType": "field",
              "value": "def-cont"
            },
            {
              "contentId": "BT-10-Procedure-Buyer",
              "contentType": "field",
              "contentCount": "2",
              "visType": "field",
              "value": "defence"
            },
            {
              "contentId": "BT-508-Procedure-Buyer",
              "contentType": "field",
              "contentCount": "2",
              "visType": "field",
              "value": "2222"
            },
            {
              "contentId": "GR-Procedure-SProvider",
              "contentType": "group",
              "contentCount": "3",
              "children": [
                {
                  "contentId": "OPT-300-Procedure-SProvider",
                  "contentType": "field",
                  "contentCount": "3",
                  "visType": "field",
                  "value": ""
                },
                {
                  "contentId": "OPT-030-Procedure-SProvider",
                  "contentType": "field",
                  "contentCount": "3",
                  "visType": "field",
                  "value": ""
                }
              ],
              "visType": "non-field",
              "visNodeId": "ND-ServiceProviderParty"
            }
          ],
          "visType": "non-field",
          "visNodeId": "ND-ContractingParty"
        }`

Physical model:

`   <cac:ContractingParty>
      <cac:Party>
         <cac:PartyIdentification>
            <cbc:ID>ORG-0001</cbc:ID>
         </cac:PartyIdentification>
      </cac:Party>
      <cac:ServiceProviderParty>
         <cac:Party>
            <cac:PartyIdentification>
               <cbc:ID/>
            </cac:PartyIdentification>
         </cac:Party>
         <cbc:ServiceTypeCode listName="organisation-role-service"/>
      </cac:ServiceProviderParty>
      <cac:ContractingPartyType>
         <cbc:PartyTypeCode listName="eforms-buyer-legal-type">body-pl</cbc:PartyTypeCode>
         <cbc:PartyTypeCode listName="eforms-buyer-legal-type">def-cont</cbc:PartyTypeCode>
      </cac:ContractingPartyType>
      <cac:ContractingActivity>
         <cbc:ActivityTypeCode listName="authority-activity">defence</cbc:ActivityTypeCode>
      </cac:ContractingActivity>
      <cbc:BuyerProfileURI>2222</cbc:BuyerProfileURI>
   </cac:ContractingParty>`

Is it because we have not completed the file correctly or the generation of repeatable fields in the generation of the physical model is not fully operational?

Thanks in advance,
Alex

**Comment by YvesJo** [comment]
Hi Alex,
Buyer is repeatable and so does cac:ContractingParty
![image](https://user-images.githubusercontent.com/42802091/218036705-599ea5a8-18fd-4e77-b8cf-2221d1cf19c0.png)
Also do not use a same organization identifier (ORG-0001) to designate organizations with "conflicting" roles as this is confusing (I see that ORG-0001 is providing services to ORG-0001).
Empty elements should be removed from the generated XML.
I wish this helps
KR
Yves

**Reply by AlexAS81** [reply]
Could you tell me which repetitive field of the eForms Notice Editor Demo form appears repetitively and correctly in the physical model when using the "Generate XML" functionality?

**Reply by rouschr** [reply]
See my answer.

**Comment by rouschr** [comment, marked as answer]
Please provide the used SDK version and the chosen notice subtype (1, 2, 3, T01, ..?) and which branch of the editor you are using. 
Did you load this XML from an existing notice or did you create it fully using an empty form?
Ideally I also need a minimal reproducible case (MRE), just the full visual model json would help a lot.
Otherwise I will have trouble reproducing it.

The visual model looks OK.
The first value is for "TED eSender" is:
```
"children": [
 { "contentId": "OPT-300-Procedure-SProvider", "contentType": "field", "contentCount": "1", 
 "visType": "field", "value": "ORG-0001" }, 
 { "contentId": "OPT-030-Procedure-SProvider", "contentType": "field", "contentCount": "1", 
 "visType": "field", "value": "ted-esen" } 
 ]
```

The second value for "Procurement Service Provider" is:
```
 { "contentId": "OPT-300-Procedure-SProvider", "contentType": "field", "contentCount": "2", 
 "visType": "field", "value": "ORG-0001" }, 
 { "contentId": "OPT-030-Procedure-SProvider", "contentType": "field", "contentCount": "2", 
 "visType": "field", "value": "serv-prov" }
```

I see that you have a problem inside **required** fields, it is possible that an invalid / incomplete form cannot be correctly transformed to XML as sometimes xpath (predicates...) depends on various required fields to work.
It is also possible this particular form has a case which was not seen yet, like a problematic xpath or something wrong in the notice type definitions.

There is a minimal "self-contained" unit test showing how nested repetition is handled, not that it relies on a dummy test SDK:
https://github.com/OP-TED/eforms-notice-editor/blob/develop/src/test/java/eu/europa/ted/eforms/noticeeditor/helper/notice/SaveNoticeRepeatableTest.java
https://github.com/OP-TED/eforms-notice-editor/tree/develop/src/test/resources/dummy-sdk/1.5.0

**Reply by AlexAS81** [reply]
The SDK version we are using is 1.5 and the notice subtype 10.

The branch we are testing is "develop".

The XML is created fully using an empty form.

I send you the visual and physical models generated.

[physicalModel.txt](https://github.com/OP-TED/eForms-SDK/files/10742539/physicalModel.txt)
[visualModel.txt](https://github.com/OP-TED/eForms-SDK/files/10742541/visualModel.txt)

Thanks in advance,
Alex

**Reply by rouschr** [reply]
I cannot find "value": "ted-esen" in the visual model but you mentioned it before, is this really the same visual model?

> The XML is created fully using an empty form.

By that you mean using the (latest develop) eforms editor demo empty form for notice subtype 10? Or a custom form?
It looks to me like there is something wrong with the json when there is repetition:
```
{
					"contentId": "GR-Procedure-LocalLegalBasisNoID",
					"contentType": "group",
					"contentCount": "1",
					"children": [{
							"contentId": "BT-01(e)-Procedure",
							"contentType": "field",
							"contentCount": "1",
							"visType": "field",
							"value": "1234567890"
						}, {
							"contentId": "BT-01(f)-Procedure",
							"contentType": "field",
							"contentCount": "1",
							"visType": "field",
							"value": "1234567890"
						}
					],
					"visType": "non-field",
					"visNodeId": "ND-LocalLegalBasisNoID"
				}, {
					"contentId": "GR-Procedure-LocalLegalBasisNoID",
					"contentType": "group",
					"contentCount": "2",
					"children": [{
							"contentId": "BT-01(e)-Procedure",
							"contentType": "field",
							"contentCount": "2",
							"visType": "field",
							"value": "0987654321"
						}, {
							"contentId": "BT-01(f)-Procedure",
							"contentType": "field",
							"contentCount": "2",
							"visType": "field",
							"value": "0987654321"
						}
					],
					"visType": "non-field",
					"visNodeId": "ND-LocalLegalBasisNoID"
				}
```
Can you try to "git pull" to the very latest commit of develop?
I would expect a json array around anything repeating like:
{A repeatable group [{item1}, {item2}, ...]}
Or:
{A repeated field: [{item1}, {item2}, ...]}
But here for example I do not see an array:
```
...
}, {
					"contentId": "BT-23-Procedure",
					"contentType": "field",
					"contentCount": "1",
					"visType": "field",
					"value": "services"
				}, NO GROUP OR NO ARRAY HERE { 
					"contentId": "BT-531-Procedure", <---- is REPEATABLE
					"contentType": "field",
					"contentCount": "1",
					"visType": "field",
					"value": "supplies" <----------- value 1
				}, {
					"contentId": "BT-531-Procedure", <---- is REPEATABLE (same kind of field)
					"contentType": "field",
					"contentCount": "3",
					"visType": "field",
					"value": "works" <-------------- value 2
				} NO END OF ARRAY HERE, {
					"contentId": "BT-27-Procedure",
					"contentType": "field",
					"contentCount": "1",
					"visType": "field",
					"value": "100"
				} ...
```
To me this looks like something is broken either in your own visual model json or in the way the current develop branch of the editor demo project builds json from the form data. After this the physical model code will not work as it expects arrays.
I will discuss this with the team.

**Reply by AlexAS81** [reply]
We have generated new physical an visual (with repeatable elements) models.

The element ND-ProcedureAdditionalCommodityClassification (BT-26) appears twice in the visual mode.

**Reply by rouschr** [reply]
> We have generated new physical an visual (with repeatable elements) models.

How did you generate the json, did you generate it yourself by modifying something or does it come exactly from the editor demo.

**Reply by AlexAS81** [reply]
Iit comes exactly from the editor demo.

**Reply by AlexAS81** [reply]
Which github branch is best to test the eForms Notice Editor application with?

**Reply by rouschr** [reply]
Currently it is the "**develop**" branch, make sure you do a `git pull` because we may update it at any time with the latest developments.
In the future the editor demo projects will have proper versioning and I hope to add more tests (maybe integration tests).
The current unit tests are not taking the front-end part into account, they test Java code in isolation.
The editor demo may be lagging a bit behind the latest SDK features, but sometimes we first work on the editor demo to test SDK features to see how usable they are in practice and later work on the SDK to implement them.
